### ISSUE TITLE
Qt5 builds do not respect EASYBUILD_OPTARCH - or any other CFLAGS, CXXFLAGS, LDFLAGS etc

### ISSUE THREAD

**Issue by casparvl** [easybuilders/easybuild-easyblocks#1670]
**Issue**

So, the issue is, Qt doesn't listen to the 'normal' CFLAGS, CXXFLAGS, and LDFLAGS.

**Description**

To be able to play around, I've dumped an env script with eb's `--dump-env-script`:

```
source Qt5-5.10.1-intel-2018b.env
```

Then, my CFLAGS are

```
[casparl@tcn900 qt-everywhere-src-5.10.1]$ echo $CFLAGS
-O2 -O2 -xAVX -axCORE-AVX2,CORE-AVX512 -ftz -fp-speculation=safe -fp-model source
```

I run the configure line that EasyBuild generates for the intel-2018b toolchain:

```
./configure -prefix /home/casparl/.local/EB_regression/RedHatEnterpriseServer7/software/Qt5/5.10.1-intel-2018b -skip qtgamepad -release  -platform linux-icc-64
```

Then, when I start building, you can see the optimization flags that are actually passed to the compiler. Again, I use the 'make' line generated by EasyBuild (for clarity, I do a serial build):

```
LD_LIBRARY_PATH=/tmp/casparl/Qt5/5.10.1/intel-2018b/qt-everywhere-src-5.10.1/qtbase/lib:/tmp/casparl/Qt5/5.10.1/intel-2018b/qt-everywhere-src-5.10.1/qtdeclarative/lib:$LD_LIBRARY_PATH  make -j 1
icpc -c -ffunction-sections -O2 -fPIC -std=c++1y -fvisibility=hidden -fvisibility-inlines-hidden -w1 -Wall -Wcheck -wd1572,873,2259,2261,3373 -D_REENTRANT -D_REENTRANT -D_REENTRANT -DQT_VERSION_STR='"5.10.1"' -DQT_VERSION_MAJOR=5 -DQT_VERSION_MINOR=10 -DQT_VERSION_PATCH=1 -DQT_BOOTSTRAPPED -DQT_NO_CAST_TO_ASCII -DQT_NO_FOREACH -DQT_NO_CAST_FROM_ASCII -DQT_NO_NARROWING_CONVERSIONS_IN_CONNECT -DQT_BUILD_BOOTSTRAP_LIB -DQT_BUILDING_QT -DQT_ASCII_CAST_WARNINGS -DQT_MOC_COMPAT -DQT_USE_QSTRINGBUILDER -DQT_DEPRECATED_WARNINGS -DQT_DISABLE_DEPRECATED_BEFORE=0x050000 -DQT_NO_EXCEPTIONS -D_LARGEFILE64_SOURCE -D_LARGEFILE_SOURCE -DQT_NO_DEBUG -I. -I../../../include -I../../../include/QtCore -I../../../include/QtCore/5.10.1 -I../../../include/QtCore/5.10.1/QtCore -I../../../include/QtXml -I../../../include/QtXml/5.10.1 -I../../../include/QtXml/5.10.1/QtXml -I../../../mkspecs/linux-icc-64 -o .obj/qlatincodec.o ../../corelib/codecs/qlatincodec.cpp
...
```
Oh dear, no optimizations other than O2. In fact, all of the flags you see there are set by Qt.

The 'Qt' way of setting those flags is, according to the 'help' of it's configure script:

```
  -sse2 ................ Use SSE2 instructions [auto]
  -sse3/-ssse3/-sse4.1/-sse4.2/-avx/-avx2/-avx512
                         Enable use of particular x86 instructions [auto]
                         Enabled ones are still subject to runtime detection.
```
I tried that (in a clean installation directory, to make sure there is no configure 'cached'), and when I search for avx2, I get for example:
```
icpc -c -O2 -std=c++1y -fvisibility=hidden -fvisibility-inlines-hidden -w1 -Wall -Wcheck -wd1572,873,2259,2261,3373 -D_REENTRANT -
D_REENTRANT -D_REENTRANT -fPIC -DQT_NO_USING_NAMESPACE -DQT_NO_FOREACH -DENABLE_PIXMAN_DRAWHELPERS -DQT_NO_NARROWING_CONVERSIONS_I
N_CONNECT -DQT_BUILD_GUI_LIB -DQT_BUILDING_QT -DQT_NO_CAST_TO_ASCII -DQT_ASCII_CAST_WARNINGS -DQT_MOC_COMPAT -DQT_USE_QSTRINGBUILD
ER -DQT_DEPRECATED_WARNINGS -DQT_DISABLE_DEPRECATED_BEFORE=0x050000 -DQT_NO_EXCEPTIONS -D_LARGEFILE64_SOURCE -D_LARGEFILE_SOURCE -
DQT_NO_DEBUG -DQT_CORE_LIB -xCORE-AVX2 -I. -I../../include -I../../include/QtGui -I../../include/QtGui/5.10.1 -I../../include/QtGu
i/5.10.1/QtGui -I../../include/QtCore/5.10.1 -I../../include/QtCore/5.10.1/QtCore -I../../include/QtCore -I.moc -isystem /hpc/sw_t
est/EB_regression/2019/software/libpng/1.6.34-GCCcore-7.3.0/include/libpng16 -I../../mkspecs/linux-icc-64 image/qimage_avx2.cpp -o
 .obj/qimage_avx2.o
```
Note: there is nothing that looks like an -xAVX2 flag. It does seem that a 'special' version of qimage is compiled that has 'something' to do with avx2. No clue what though. Also, only about 3 functions so far (5 minutes into my compilation) seem to have 'avx2' in their name - whatever that means...

Qt5 DOES allow for setting e.g. something like CFLAGS, but it doesn't listen to the CFLAGS, CXXFLAGS and LDFLAGS environment variables. Instead, the config says this:

```
It is also possible to manipulate any QMAKE_* variable, to amend the values
from the mkspec for the build of Qt itself, e.g., QMAKE_CXXFLAGS+=-g3.
```

If I try passing the $CXXFLAGS as QMAKE_CXXFLAGS:
```
[casparl@tcn900 qt-everywhere-src-5.10.1]$ ./configure -prefix /home/casparl/.local/EB_regression/RedHatEnterpriseServer7/software/Qt5/5.10.1-intel-2018b -skip qtgamepad -release  -platform linux-icc-64 QMAKE_CXXFLAGS+="$CXXFLAGS"
```
I get for example:
```
icpc -c -O2 -O2 -xAVX -axCORE-AVX2,CORE-AVX512 -ftz -fp-speculation=safe -fp-model source -ffunction-sections -O2 -fPIC -std=c++1y -fvisibility=hidden -fvisibility-inlines-hidden -w1 -Wall -Wcheck -wd1572,873,2259,2261,3373 -D_REENTRANT -D_REENTRANT -D_REENTRANT -DQT_VERSION_STR='"5.10.1"' -DQT_VERSION_MAJOR=5 -DQT_VERSION_MINOR=10 -DQT_VERSION_PATCH=1 -DQT_BOOTSTRAPPED -DQT_NO_CAST_TO_ASCII -DQT_NO_FOREACH -DQT_NO_CAST_FROM_ASCII -DQT_NO_NARROWING_CONVERSIONS_IN_CONNECT -DQT_BUILD_BOOTSTRAP_LIB -DQT_BUILDING_QT -DQT_ASCII_CAST_WARNINGS -DQT_MOC_COMPAT -DQT_USE_QSTRINGBUILDER -DQT_DEPRECATED_WARNINGS -DQT_DISABLE_DEPRECATED_BEFORE=0x050000 -DQT_NO_EXCEPTIONS -D_LARGEFILE64_SOURCE -D_LARGEFILE_SOURCE -DQT_NO_DEBUG -I. -I../../../include -I../../../include/QtCore -I../../../include/QtCore/5.10.1 -I../../../include/QtCore/5.10.1/QtCore -I../../../include/QtXml -I../../../include/QtXml/5.10.1 -I../../../include/QtXml/5.10.1/QtXml -I../../../mkspecs/linux-icc-64 -o .obj/qutfcodec.o ../../corelib/codecs/qutfcodec.cpp
../../corelib/codecs/qutfcodec.cpp(522): (col. 1) remark: _ZN6QUtf1618convertFromUnicodeEPK5QChariPN10QTextCodec14ConverterStateE14DataEndianness has been targeted for automatic cpu dispatch
../../corelib/codecs/qutfcodec.cpp(641): (col. 1) remark: _ZN6QUtf3218convertFromUnicodeEPK5QChariPN10QTextCodec14ConverterStateE14DataEndianness has been targeted for automatic cpu dispatch
../../corelib/codecs/qutfcodec.cpp(800): (col. 1) remark: _ZNK11QUtf16Codec18convertFromUnicodeEPK5QChariPN10QTextCodec14ConverterStateE has been targeted for automatic cpu dispatch
../../corelib/codecs/qutfcodec.cpp(861): (col. 1) remark: _ZNK11QUtf32Codec18convertFromUnicodeEPK5QChariPN10QTextCodec14ConverterStateE has been targeted for automatic cpu dispatch
```
This is the behaviour I would actually expect! And, I would expect the EasyBlock for Qt to translate things like CFLAGS, CXXFLAGS and LDFLAGS to their Qt-equivalents. (see https://doc.qt.io/qt-5/qmake-variable-reference.html)

That's where I would like to start a discussion. 

**First question**

Should we 'translate' CFLAGS, CXXFLAGS, and LDFLAGS into Qt environment variables such as QMAKE_CXXFLAG? (this 'bypasses' if you will the standard -avx2 flag that you can already give to Qt5's configure, but then again, that flag alone doesn't seem to REALLY optimize in a way we want) 

**Second question**

If so, which Qt variables should we set? E.g. there is

- QMAKE_CXXFLAGS
- QMAKE_CXXFLAGS_DEBUG
- QMAKE_CXXFLAGS_RELEASE
- QMAKE_CXXFLAGS_SHLIB
- QMAKE_CXXFLAGS_THREAD
- QMAKE_CXXFLAGS_WARN_OFF
- QMAKE_CXXFLAGS_WARN_ON

Similar variations exist for QMAKE_CFLAGS and QMAKE_LFLAGS (Qt variant of LDFLAGS).

Should we set all of these to CXXFLAGS? And should we do QMAKE_CXXFLAGS = "$CXXFLAGS", or QMAKE_CXXFLAGS += "$CXXFLAGS"?

**Comment by casparvl** [comment]
Logging from the chat...:

boegel [1:08 PM]
@Caspar van Leeuwen passing down `$CXXFLAGS` & co via `$QMAKE_CXXFLAGS` makes most sense to me
I do wonder how many Qt* easyconfigs this is going to affect/break... :confused:

Caspar van Leeuwen [1:18 PM]
good question. Since the Qt EasyConfigs themselves don’t really add CFLAGS (well, I’m planning to because i have issues with -ipo for intel-2018b, so I wan’t to pass -no-ipo…) the ‘only’ thing this would change is the optimization. However, there is a second choice to make: Qt supports constructs like `QMAKE_CXXFLAGS += "$CXXFLAGS"?`, or `QMAKE_CXXFLAGS = "$CXXFLAGS",`. I’m not sure if in practice there will be a difference, since I think the environment variables are empty at the time the configure runs, but I don’t know if/how it affects Qt’s .pro files that add to these variables…
Anyway, it is definitely the easiest approach I think, easier than trying to do things like (pseudo-code…) `if EASYBUILD_OPTARCHS =~ avx2; then configopts += -avx2`.
I’ll give a try to adapt the EasyBlock. Is there an ‘EasyBuild-way’ of setting environment variables (e.g. so that eb --dump-env-script would also dump them?)

boegel [1:21 PM]
@Caspar van Leeuwen Use the `setvar` function from `easybuild.tools.environment`, it's used in other easyblocks too

Caspar van Leeuwen [1:22 PM]
ok great, I’ll give it a try

**Comment by casparvl** [comment]
One question is still _which_ QMAKE_CXXFLAGS* to set. It is my understanding that QMAKE_CXXFLAGS applies to all builds, and e.g. QMAKE_CXXFLAGS_DEBUG is there to fine-tune flags for debug builds. Thus, to me, it makes sense to just set QMAKE_CXXFLAGS="$CXXFLAGS", so that it applies to all builds. 

Note that Qt allows something like QMAKE_CXXFLAGS+="$CXXFLAGS" to be specified on the configure line. I guess this adds to any existing QMAKE_CXXFLAGS already specified. Currently, I'm exploring the approach of just setting QMAKE_CXXFLAGS as environment variable, rather than specifying it as an argument to `configure`.

**Comment by casparvl** [comment]
Qt's `configure` doesn't pick up on `QMAKE_CXXFLAGS` as environment variable. It _has_ to be specified as argument on configure command line. Now trying that approach...

**Comment by casparvl** [comment]
I have inserted this in the beginning of the configure_step of the qt.py EasyBlock:

```
        # Qt (at least Qt5) doesn't respect CFLAGS, CXXFLAGS and LDFLAGS, but has equivalent flags,
        # e.g. QMAKE_CFLAGS. Since EasyBuild relies e.g. for EASYBUILD_OPTARCH on CLFAGS, we need to
        # set the equivalent QMAKE_* flags as configure arguments.
        # See https://github.com/easybuilders/easybuild-easyblocks/issues/1670
        # For compilers, we replace QMAKE_CC (otherwise, you get e.g. QMAKE_CC="g++ g++", which fails)
        env_to_options = {
            'CC': 'QMAKE_CC',
            'CXX': 'QMAKE_CXX',
        }
        for env_name, option in env_to_options.items():
            value = os.getenv(env_name)
            if value is not None:
                self.cfg.update('configopts', '%s="%s"' % (option, value))
        # For FLAGS, we add to existing flags (e.g. those set in Qt's .pro-files).
        env_to_options = {
            'CFLAGS': 'QMAKE_CFLAGS',
            'CXXFLAGS': 'QMAKE_CXXFLAGS',
            'LDFLAGS': 'QMAKE_LFLAGS',
        }
        for env_name, option in env_to_options.items():
            value = os.getenv(env_name)
            if value is not None:
                self.cfg.update('configopts', '%s+="%s"' % (option, value))
```

That has the desired effect of passing the CFLAGS, CXXFLAGS, and LDFLAGS on the compile/link lines.

I did encounter another problem regarding AVX-512 optimization (see https://github.com/easybuilders/easybuild-easyconfigs/issues/8003). I do however feel that since there is a proper workaround for this (switch off AVX-512 support for Qt5, which is not a problem in my view since the current EasyBlock will compile without ANY optimization), this does not have to block integration of my adaptation to qt.py. I'll figure out how to do a proper pull request (EasyBuild's github integration only works for EasyConfigs, right?) to the official EasyBlock repo.

**Comment by casparvl** [comment]
@boegel asked me to post a diff. The previous post essentially contains that, but to avoid any possible confusion:

```
[casparl@int1 ~]$ diff /hpc/eb/RedHatEnterpriseServer7/EasyBuild/3.8.1/lib/python2.7/site-packages/easybuild_easyblocks-3.8.1-py2.7.egg/easybuild/easyblocks/q/qt.py easyblocks-surfsara/qt.py
41d40
<
61a61,84
>
>         # Qt (at least Qt5) doesn't respect CFLAGS, CXXFLAGS and LDFLAGS, but has equivalent flags,
>         # e.g. QMAKE_CFLAGS. Since EasyBuild relies e.g. for EASYBUILD_OPTARCH on CLFAGS, we need to
>         # set the equivalent QMAKE_* flags as configure arguments.
>         # See https://github.com/easybuilders/easybuild-easyblocks/issues/1670
>         # For compilers, we replace QMAKE_CC (otherwise, you get e.g. QMAKE_CC="g++ g++", which fails)
>         env_to_options = {
>             'CC': 'QMAKE_CC',
>             'CXX': 'QMAKE_CXX',
>         }
>         for env_name, option in env_to_options.items():
>             value = os.getenv(env_name)
>             if value is not None:
>                 self.cfg.update('configopts', '%s="%s"' % (option, value))
>         # For FLAGS, we add to existing flags (e.g. those set in Qt's .pro-files).
>         env_to_options = {
>             'CFLAGS': 'QMAKE_CFLAGS',
>             'CXXFLAGS': 'QMAKE_CXXFLAGS',
>             'LDFLAGS': 'QMAKE_LFLAGS',
>         }
>         for env_name, option in env_to_options.items():
>             value = os.getenv(env_name)
>             if value is not None:
>                 self.cfg.update('configopts', '%s+="%s"' % (option, value))
```

**Comment by boegel** [comment]
I'm looking into making the proposed changes to the `Qt` easyblock for (last-minute) inclusion into EasyBuild v3.9.0.

One thing I ran into is that these `configure` options are only supported by `Qt5`, not for `Qt` 4.x, since trying to use them results in the `configure` command failing because `QMAKE_*` is not known.

It'll be important to also take into account the issue reported in https://github.com/easybuilders/easybuild-easyconfigs/issues/8003, i.e. we'll have to disable `optarch` for some `Qt5` easyconfigs that use particular versions of the `intel` toolchain...

**Comment by boegel** [comment]
Turns out these `QMAKE_*` configure options are only supported for the latest versions of `Qt5`...

With `Qt5` 5.7.0 for example, `configure` fails with:

```
QMAKE_CC=icc: unknown argument
QMAKE_CFLAGS+=-O2 -xHost -ftz -fp-speculation=safe -fp-model source: unknown argument
QMAKE_CXX=icpc: unknown argument
QMAKE_CXXFLAGS+=-O2 -xHost -ftz -fp-speculation=safe -fp-model source: unknown argument
...
```

**Comment by boegel** [comment]
fixed in #1677 for recent `Qt5` versions (5.8.0 & newer), so closing this...

Thanks for reporting this @casparvl!
